File: template-parts/code-library/snippets/go/rate_limiter.go
// Package ratelimiter implements a token bucket rate limiter for Go.
// Purpose: Control the rate of operations to prevent overwhelming downstream services.
// Usage:
//   limiter := ratelimiter.New(100, time.Second) // 100 requests per second
//   if err := limiter.Wait(ctx); err != nil { return err }
//   doWork()
//
// Dependencies: standard library only
package ratelimiter

import (
	"context"
	"errors"
	"sync"
	"time"
)

// ErrRateLimited is returned when the rate limit prevents a request.
var ErrRateLimited = errors.New("rate limit exceeded")

// Limiter implements a token bucket rate limiter.
type Limiter struct {
	mu       sync.Mutex
	rate     float64         // tokens per second
	capacity int             // max tokens (bucket size)
	tokens   float64         // current available tokens
	lastTime time.Time       // last token refill time

	// Optional channel for async notification
	notifyCh chan struct{}
}

// New creates a new rate limiter.
// rate specifies tokens per second, capacity is the max bucket size.
func New(rate float64, capacity int) *Limiter {
	return &Limiter{
		rate:       rate,
		capacity:   capacity,
		tokens:     float64(capacity),
		lastTime:   time.Now(),
		notifyCh:   make(chan struct{}, 1),
	}
}

// NewWithNotify creates a limiter that notifies via channel when tokens are available.
func NewWithNotify(rate float64, capacity int) *Limiter {
	limiter := New(rate, capacity)
	limiter.notifyCh = make(chan struct{}, capacity)
	return limiter
}

// Wait blocks until a token is available or context is cancelled.
func (l *Limiter) Wait(ctx context.Context) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	default:
	}

	if err := l.reserve(1); err != nil {
		return err
	}

	// Wait for token if none available
	if l.tokens < 0 {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(l.estimateWaitTime()):
			return nil
		}
	}
	return nil
}

// Try acquires n tokens without blocking.
// Returns nil on success, ErrRateLimited if not enough tokens.
func (l *Limiter) Try(n int) error {
	return l.reserve(n)
}

// reserve attempts to reserve n tokens.
func (l *Limiter) reserve(n int) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	l.refill()

	if float64(n) > l.tokens {
		return ErrRateLimited
	}

	l.tokens -= float64(n)
	return nil
}

// refill adds tokens based on elapsed time since last refill.
func (l *Limiter) refill() {
	now := time.Now()
	elapsed := now.Sub(l.lastTime).Seconds()
	l.lastTime = now

	// Add tokens based on rate and elapsed time
	l.tokens += elapsed * l.rate
	if l.tokens > float64(l.capacity) {
		l.tokens = float64(l.capacity)
	}
}

// estimateWaitTime estimates how long until a token is available.
func (l *Limiter) estimateWaitTime() time.Duration {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.tokens >= 0 {
		return 0
	}
	return time.Duration((-l.tokens / l.rate) * float64(time.Second))
}

// Tokens returns the current number of available tokens.
func (l *Limiter) Tokens() float64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.refill()
	return l.tokens
}

// Capacity returns the maximum token capacity.
func (l *Limiter) Capacity() int {
	return l.capacity
}

// Rate returns the token generation rate per second.
func (l *Limiter) Rate() float64 {
	return l.rate
}

// Reset restores the limiter to full capacity.
func (l *Limiter) Reset() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.tokens = float64(l.capacity)
	l.lastTime = time.Now()
}

// Burst attempts to consume up to n tokens instantly.
// Returns the number of tokens actually consumed.
func (l *Limiter) Burst(n int) int {
	l.mu.Lock()
	defer l.mu.Unlock()

	l.refill()

	available := int(l.tokens)
	if available > n {
		available = n
	}
	if available > l.capacity {
		available = l.capacity
	}

	l.tokens -= float64(available)
	return available
}

File: template-parts/code-library/snippets/python/rate_limiter.py
"""Token bucket rate limiter for Python.

Controls the rate of operations to prevent overwhelming downstream services.

Usage:
    limiter = TokenBucketRateLimiter(rate=100, capacity=100)  # 100 per second
    await limiter.acquire()
    await do_work()

    # Or sync version
    limiter = SyncTokenBucketRateLimiter(rate=10, capacity=10)
    limiter.acquire()
    do_work()
"""

from __future__ import annotations

import asyncio
import threading
import time
from typing import Final


class TokenBucketRateLimiter:
    """Async token bucket rate limiter.

    Args:
        rate: Number of tokens added per second.
        capacity: Maximum number of tokens in the bucket.
    """

    def __init__(self, rate: float, capacity: int) -> None:
        if rate <= 0:
            raise ValueError("rate must be positive")
        if capacity <= 0:
            raise ValueError("capacity must be positive")

        self._rate: Final[float] = rate
        self._capacity: Final[int] = capacity
        self._tokens: float = float(capacity)
        self._last_update: float = time.monotonic()
        self._lock: asyncio.Lock = asyncio.Lock()

    async def acquire(self, tokens: int = 1) -> None:
        """Acquire tokens, waiting if necessary.

        Args:
            tokens: Number of tokens to acquire.
        """
        await self._wait_for_tokens(tokens)

    async def _wait_for_tokens(self, tokens: int) -> None:
        async with self._lock:
            await self._refill()

            if self._tokens >= tokens:
                self._tokens -= tokens
                return

            # Calculate wait time
            needed = tokens - self._tokens
            wait_time = needed / self._rate

        await asyncio.sleep(wait_time)

        async with self._lock:
            await self._refill()
            self._tokens -= tokens

    async def _refill(self) -> None:
        """Refill tokens based on elapsed time."""
        now = time.monotonic()
        elapsed = now - self._last_update
        self._last_update = now

        self._tokens += elapsed * self._rate
        if self._tokens > self._capacity:
            self._tokens = float(self._capacity)

    async def try_acquire(self, tokens: int = 1) -> bool:
        """Try to acquire tokens without blocking.

        Returns:
            True if tokens were acquired, False otherwise.
        """
        async with self._lock:
            await self._refill()

            if self._tokens >= tokens:
                self._tokens -= tokens
                return True
            return False

    @property
    def tokens(self) -> float:
        """Current number of available tokens."""
        return self._tokens

    @property
    def capacity(self) -> int:
        """Maximum bucket capacity."""
        return self._capacity

    @property
    def rate(self) -> float:
        """Token generation rate per second."""
        return self._rate


class SyncTokenBucketRateLimiter:
    """Synchronous token bucket rate limiter.

    Args:
        rate: Number of tokens added per second.
        capacity: Maximum number of tokens in the bucket.
    """

    def __init__(self, rate: float, capacity: int) -> None:
        if rate <= 0:
            raise ValueError("rate must be positive")
        if capacity <= 0:
            raise ValueError("capacity must be positive")

        self._rate = rate
        self._capacity = capacity
        self._tokens = float(capacity)
        self._last_update = time.monotonic()
        self._lock = threading.Lock()

    def acquire(self, tokens: int = 1) -> None:
        """Acquire tokens, blocking until available.

        Args:
            tokens: Number of tokens to acquire.
        """
        with self._lock:
            self._refill_locked()

            if self._tokens >= tokens:
                self._tokens -= tokens
                return

            # Calculate wait time
            needed = tokens - self._tokens
            wait_time = needed / self._rate

        time.sleep(wait_time)

        with self._lock:
            self._refill_locked()
            self._tokens -= tokens

    def _refill_locked(self) -> None:
        """Refill tokens based on elapsed time. Must be called with lock held."""
        now = time.monotonic()
        elapsed = now - self._last_update
        self._last_update = now

        self._tokens += elapsed * self._rate
        if self._tokens > self._capacity:
            self._tokens = float(self._capacity)

    def try_acquire(self, tokens: int = 1) -> bool:
        """Try to acquire tokens without blocking.

        Returns:
            True if tokens were acquired, False otherwise.
        """
        with self._lock:
            self._refill_locked()

            if self._tokens >= tokens:
                self._tokens -= tokens
                return True
            return False

    @property
    def tokens(self) -> float:
        """Current number of available tokens."""
        with self._lock:
            self._refill_locked()
            return self._tokens

    def reset(self) -> None:
        """Reset the limiter to full capacity."""
        with self._lock:
            self._tokens = float(self._capacity)
            self._last_update = time.monotonic()


class SlidingWindowRateLimiter:
    """Sliding window rate limiter.

    Tracks requests in a sliding time window for more precise rate limiting.

    Args:
        rate: Maximum number of requests allowed in the window.
        window_seconds: Size of the sliding window in seconds.
    """

    def __init__(self, rate: int, window_seconds: float) -> None:
        self._rate = rate
        self._window_seconds = window_seconds
        self._requests: list[float] = []
        self._lock = threading.Lock()

    async def acquire(self) -> None:
        """Acquire a slot, waiting if the rate limit is exceeded."""
        while True:
            if await self._try_acquire_async():
                return
            await asyncio.sleep(0.01)

    def acquire_sync(self) -> None:
        """Synchronous acquire."""
        while True:
            if self._try_acquire_sync():
                return
            time.sleep(0.01)

    async def _try_acquire_async(self) -> bool:
        async with self._lock:
            return self._try_acquire_locked()

    def _try_acquire_sync(self) -> bool:
        with self._lock:
            return self._try_acquire_locked()

    def _try_acquire_locked(self) -> bool:
        now = time.monotonic()
        cutoff = now - self._window_seconds

        # Remove expired requests
        self._requests = [t for t in self._requests if t > cutoff]

        if len(self._requests) < self._rate:
            self._requests.append(now)
            return True
        return False
